lib/ejs.js

```diff
@@ -721,75 +712,97 @@ Template.prototype = {
 
     newLineCount = (line.split('\n').length - 1);
 
-    switch (line) {
-    case '<' + d:
-    case '<' + d + '_':
-      this.mode = Template.modes.EVAL;
-      break;
-    case '<' + d + '=':
-      this.mode = Template.modes.ESCAPED;
-      break;
-    case '<' + d + '-':
-      this.mode = Template.modes.RAW;
-      break;
-    case '<' + d + '#':
-      this.mode = Template.modes.COMMENT;
-      break;
-    case '<' + d + d:
-      this.mode = Template.modes.LITERAL;
-      this.source += '    ; __append("' + line.replace('<' + d + d, '<' + d) + '")' + '\n';
-      break;
-    case d + d + '>':
-      this.mode = Template.modes.LITERAL;
-      this.source += '    ; __append("' + line.replace(d + d + '>', d + '>') + '")' + '\n';
-      break;
-    case d + '>':
-    case '-' + d + '>':
-    case '_' + d + '>':
-      if (this.mode == Template.modes.LITERAL) {
-        this._addOutput(line);
+    var isHandled;
+    if (line[0] == '<') {
+      isHandled = true;
+      switch (line) {
+      case '<' + d:
+      case '<' + d + '_':
+        this.mode = Template.modes.EVAL;
+        break;
+      case '<' + d + '=':
+        this.mode = Template.modes.ESCAPED;
+        break;
+      case '<' + d + '-':
+        this.mode = Template.modes.RAW;
+        break;
+      case '<' + d + '#':
+        this.mode = Template.modes.COMMENT;
+        break;
+      case '<' + d + d:
+        this.mode = Template.modes.LITERAL;
+        this.source += '    ; __append("' + line.replace('<' + d + d, '<' + d) + '")' + '\n';
+        break;
+      default:
+        isHandled = false;
       }
-
-      this.mode = null;
-      this.truncate = line.indexOf('-') === 0 || line.indexOf('_') === 0;
-      break;
-    default:
-        // In script mode, depends on type of tag
-      if (this.mode) {
-          // If '//' is found without a line break, add a line break.
-        switch (this.mode) {
-        case Template.modes.EVAL:
-        case Template.modes.ESCAPED:
-        case Template.modes.RAW:
-          if (line.lastIndexOf('//') > line.lastIndexOf('\n')) {
-            line += '\n';
-          }
-        }
-        switch (this.mode) {
-            // Just executing code
-        case Template.modes.EVAL:
-          this.source += '    ; ' + line + '\n';
-          break;
-            // Exec, esc, and output
-        case Template.modes.ESCAPED:
-          this.source += '    ; __append(escapeFn(' + stripSemi(line) + '))' + '\n';
-          break;
-            // Exec and output
-        case Template.modes.RAW:
-          this.source += '    ; __append(' + stripSemi(line) + ')' + '\n';
-          break;
-        case Template.modes.COMMENT:
-              // Do nothing
+      if (isHandled && this.mode != Template.modes.LITERAL) {
+        if (this.isInTag)
```

Comment:
Minor quibble: `isInTag` implies a Boolean value. I can see you're leaning on the truthy/falsy in JS here, and letting this property do double-duty by holding an actual value -- but ideally if the property name looks Boolean (e.g., `isSomething`, `hasSomething`, etc.) it should actually be a Bool.